### ISSUE TITLE
Support multiple installs per game

### ISSUE THREAD

**Issue by Ortham** [loot/loot#1850]
LOOT currently supports Steam, GOG, EGS and standalone installs of its supported games (though the last only applies to Morrowind and Oblivion, which were not originally released on Steam). However, if you've got multiple copies of a game installed, it will only pick one and you need to manually add more games to LOOT's settings to cover the others.

It would be more useful in such cases if LOOT could find all installed copies of a game and create a settings entry for each one. To differentiate entries, it would make sense to name them something like `<game name>[ - <store>][ - <number>]`, where `<number>` is an incrementing number to distinguish multiple copies of a game (e.g. if there are two gog installs or a Steam install wasn't moved properly). 

To do this, LOOT will need to tell where a game was installed from. This is particularly important for games found through the Registry, as all games have a common Registry entry that may be written by multiple install sources (e.g. Enderal SE has a non-store-specific Registry key that is written by both the Steam and GOG versions), but some versions use different local folders to store their load order (e.g. Skyrim SE uses different folders for Steam, GOG and EGS).

Here's a table of the supported games and their sources. I've used Standalone to represent the non-store-specific Registry key, even if the game in question has only ever been obtainable from digital storefronts.

| Game | Sources |
|-------|---------|
| Morrowind | Standalone, Steam, GOG, MS Store
| Oblivion | Standalone, Steam, GOG, MS Store
| Nehrim | Standalone, Steam, GOG
| Skyrim | Standalone, Steam
| Enderal | Standalone, Steam
| Skyrim SE | Standalone, Steam, GOG, MS Store, EGS
| Enderal SE | Standalone, Steam, GOG
| Skyrim VR | Standalone, Steam
| Fallout 3 | Standalone, Steam, GOG, MS Store, EGS
| Fallout New Vegas | Standalone, Steam, GOG, MS Store
| Fallout 4 | Standalone, Steam, MS Store
| Fallout 4 VR | Standalone, Steam

When it comes to detecting if an install came from GOG, LOOT currently looks for `Galaxy64.dll` in the game install path when disambiguating Enderal SE, but that's not consistently available. What does seem to work is looking for a `goggame-<game id>.ico` file, where `<game id>` is the GOG ID number for the game, the same as is used in the GOG-specific Registry keys. I've verified this works for Morrowind, Oblivion, Nehrim, Skyrim SE and Enderal SE - I don't have Fallout 3 or Fallout New Vegas on GOG to check them.

I don't think it's necessary to disambiguate MS Store and EGS installs from each other or others, as detecting those installs involves special logic that wouldn't have been necessary if they wrote to the non-store-specific Registry keys. However, Morrowind, Oblivion and Fallout 3 were (and maybe still are?) available as physical and digital copies not tied to any digital distribution platform, so for them it would be good to tell if such an install is present. I don't have Morrowind on Steam to check if it's got any obviously Steam-specific files though.

So, outstanding questions:

1. Does a GOG install of Fallout 3 have a `goggame-1248282609.ico` and/or `goggame-1454315831.ico` file in its install folder? 
2. Does a GOG install of Fallout New Vegas have a `goggame-1312824873.ico` and/or `goggame-1454587428.ico` file in its install folder?
3. Does a Steam install of Morrowind have any obviously Steam-specific files in its install folder, e.g. a `steam_api.dll` file?

**Comment by sibir-ine** [comment]
Here is a full list of files Steam installs for Morrowind by default: [File_List.txt](https://github.com/loot/loot/files/11012648/File_List.txt)

Doesn't look like there's anything Steam specific, not even `steam_api.dll`, & I don't have a standalone version to compare. However, running the launcher does create `steam_autocloud.vdf` & `Saves\steam_autocloud.vdf` which are used for the Steam Cloud game save backup feature. The files are created regardless of whether the feature is enabled for Morrowind or globally, & they aren't removed when verifying game file integrity. Therefore, unless we can find a better alternative, they should still be usable.

**Comment by Ortham** [comment]
I saw that Fallout 3 and Fallout New Vegas are currently on sale on GOG, so I bought and downloaded them, and added their info to the table above.

**Comment by sibir-ine** [comment]
> Also, all the EGS games I've got installed have a `.egstore` directory (including Fallout 3), if they ever need disambiguating. I don't currently have any games from the MS Store installed to check them.

I imagine we could just use the fact that they require special methods to locate them to also identify them as EGS or MS games. Just in case, this is a full list of files installed for MS Morrowind relative to `The Elder Scrolls III- Morrowind (PC)/`: [File_List_MS.txt](https://github.com/loot/loot/files/11025002/File_List_MS.txt). `MicrosoftGame.Config` or `appxmanifest.xml` would probably make the most sense to use.

**Comment by Ortham** [comment]
I've implemented this in the `overhaul-game-detection` branch - it was a pretty significant change, adding about 2000 lines of code.

As a result of the changes, game settings have lost the "Is instance of base game" and the registry keys settings, as they're now obsolete: game detection is now aware of Nehrim and Enderal as games in their own right, and the registry keys are now hardcoded for the supported games (it made sense to split them up into GOG, Steam and generic keys, and if another total conversion with its own registry keys comes along, we can add support and until then it probably makes more sense for users to just supply the game's install path anyway). It wasn't strictly speaking necessary to remove the ability to provide registry keys, but I didn't think it was worth keeping - I've never heard of anyone modifying them.

Screenshot of the games dropdown showing my installed copies of Skyrim SE:

![image](https://user-images.githubusercontent.com/2430738/230735464-810be572-bc60-4a6b-b44d-addd5d11ce2b.png)

and here's the settings dialog with those settings removed:

![image](https://user-images.githubusercontent.com/2430738/230735482-f2ba5fb5-ef52-46ad-a19f-c22d40e37689.png)

**Comment by Ortham** [comment]
@pStyl3: Do you want to give the branch a spin? (I'm not sure if you've got multiple copies or if your suggestion in Discord was with others in mind.)

**Comment by pStyl3** [comment]
Testing with `loot_0.19.1-53-gcc5c1b9_overhaul-game-detection-win64.7z`

Overall this build seems to be working nicely, with only a few minor things I noticed.

Here's a screenshot of my games dropdown:

![grafik](https://user-images.githubusercontent.com/10406657/230736666-94c4267f-955c-413e-9125-85c48f02324f.png)

Those are valid matches, as I have them all installed. That being said, I've got Fallout 3 from GOG, shouldn't the entry be listed as `Fallout 3 (GOG)` in the games dropdown menu?

In the settings menu, the `Install Path` of Fallout 3 is also listed as `d:\gog\fallout 3`, so in contrary to the other games this field writes it's content with lower case letters only. The `Install Path` for my SSE GOG is `D:\GOG\Skyrim Anniversary Edition` for example.


`LOOTDebugLog.txt` includes the following lines:
```
[20:02:52.902119] [error]: Error while checking if game "TES V: Skyrim Special Edition" is installed through the Epic Games Store: directory_iterator::directory_iterator: Das System kann den angegebenen Pfad nicht finden.: "C:\ProgramData\Epic\EpicGamesLauncher\Data\Manifests"
[20:02:52.905556] [error]: Error while checking if game "Fallout 3" is installed through the Epic Games Store: directory_iterator::directory_iterator: Das System kann den angegebenen Pfad nicht finden.: "C:\ProgramData\Epic\EpicGamesLauncher\Data\Manifests"
```
Shouldn't these be warnings, rather than errors?

**Comment by Ortham** [comment]
> Those are valid matches, as I have them all installed. That being said, I've got Fallout 3 from GOG, shouldn't the entry be listed as `Fallout 3 (GOG)` in the games dropdown menu?

No, LOOT only adds the install source suffix if it finds more than one copy from different sources. In this case there's only the GOG copy, so it's no different from all the other games you've got other than Skyrim SE.

> In the settings menu, the `Install Path` of Fallout 3 is also listed as `d:\gog\fallout 3`, so in contrary to the other games this field writes it's content with lower case letters only. The `Install Path` for my SSE GOG is `D:\GOG\Skyrim Anniversary Edition` for example.

Is the path for FO3 in the Registry in lowercase too? LOOT doesn't do any lowercasing of paths.

> Shouldn't these be warnings, rather than errors?

Ah, I'm missing a check to see if that directory exists before trying to iterate over its contents, so it should be an error but you shouldn't be seeing it.

**Comment by pStyl3** [comment]
> No, LOOT only adds the install source suffix if it finds more than one copy from different sources. In this case there's only the GOG copy, so it's no different from all the other games you've got other than Skyrim SE.

Okay, makes sense.

> Is the path for FO3 in the Registry in lowercase too? LOOT doesn't do any lowercasing of paths.

Yes it is, so everything seems to be okay.

**Comment by sibir-ine** [comment]
Looks good to me, these are the only duplicate games I have:

![image](https://user-images.githubusercontent.com/69141501/230828676-66ce1739-f604-4528-bb57-997172f775c5.png)
